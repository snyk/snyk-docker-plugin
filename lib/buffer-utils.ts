import * as crypto from "crypto";
import { HashAlgorithm } from "./types";

const HASH_ENCODING = "hex";

export async function bufferToSha1(buffer: Buffer): Promise<string> {
  const hash = crypto.createHash(HashAlgorithm.Sha1);
  const chunkSize = 100 * 1024 * 1024; // 100 MB

  return new Promise((resolve, reject) => {
    try {
      for (let offset = 0; offset < buffer.length; offset += chunkSize) {
        const end = Math.min(offset + chunkSize, buffer.length);
        const chunk = buffer.slice(offset, end);
        hash.update(chunk);
      }

      const digest = hash.digest(HASH_ENCODING);
      resolve(digest);
    } catch (err) {
      reject(err);
    }
  });
}
