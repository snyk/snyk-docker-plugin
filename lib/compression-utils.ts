import { Decompress as ZstdDecompress } from "fzstd";

/**
 * Decompresses zstd-compressed data from a buffer.
 *
 * This is a synchronous buffer-to-buffer decompression utility.
 * For streaming zstd decompression, use the decompressMaybe transform stream.
 *
 * @param compressed Buffer containing zstd-compressed data
 * @returns Decompressed data as a Buffer
 * @throws Error if decompression fails
 */
export function decompressZstd(compressed: Buffer): Buffer {
  const chunks: Buffer[] = [];

  try {
    const decompressor = new ZstdDecompress((data: Uint8Array) => {
      chunks.push(Buffer.from(data));
    });

    decompressor.push(new Uint8Array(compressed), true);

    return Buffer.concat(chunks);
  } catch (error) {
    throw new Error(
      `Zstd decompression failed: ${
        error instanceof Error ? error.message : String(error)
      }`,
    );
  }
}
