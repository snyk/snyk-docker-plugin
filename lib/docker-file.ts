import * as fs from 'fs';
import * as dockerFileParser from 'docker-file-parser';

export { getBaseImageName };

async function getBaseImageName(targetFile?: string):
  Promise<string|undefined> {

  if (!targetFile) {
    return undefined;
  }

  const contents = await readFile(targetFile);
  const commands = dockerFileParser.parse(contents);
  const fromCommands = commands.filter(command => command.name === 'FROM');
  const finalFrom = fromCommands.pop();

  return finalFrom ? String(finalFrom.args) : undefined;
}

async function readFile(path: string) {
  return new Promise((resolve, reject) => {
    fs.readFile(path, 'utf8', (err, data) => {
      return err ? reject(err) : resolve(data);
    });
  }) as Promise<string>;
}
