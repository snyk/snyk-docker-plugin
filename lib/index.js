var subProcess = require('./sub-process');
var fetchSnykDockerAnalyzer = require('./fetch-snyk-docker-analyzer');
var debug = require('debug')('snyk');

module.exports = {
  inspect: inspect,
};

function inspect(root, targetFile, options) {
  const targetImage = root;
  return fetchSnykDockerAnalyzer()
  .then(function (analyzerBinaryPath) {
    return Promise.all([
      getRuntime(),
      getDependencies(analyzerBinaryPath, targetImage),
    ])
    .then(function (result) {
      var metadata = {
        name: 'snyk-docker-plugin',
        runtime: result[0],
        packageManager: result[1].packageManager,
      };
      var package = result[1].package;
      return {
        plugin: metadata,
        package: package,
      };
    });
  })
}

function getRuntime() {
  return subProcess.execute('docker', ['version'])
  .then(function (output) {
    versionMatch = /Version:\s+(.*)\n/.exec(output);
    if (versionMatch) {
      return 'docker ' + versionMatch[1]
    }
    return undefined;
  });
}

function handleCommonErrors(error, targetImage) {
  if (error.indexOf('command not found') !== -1) {
    throw new Error('Snyk docker CLI was not found')
  }
  if (error.indexOf('Cannot connect to the Docker daemon') !== -1) {
    throw new Error('Cannot connect to the Docker daemon. Is the docker'
    + ' daemon running?')
  }
  if ((error.indexOf('Error loading image from docker engine') !== -1) ||
      (error.indexOf('Error performing image analysis') !== -1)) {
    throw new Error('Docker image was not found: ' + targetImage)
  }
  if (error.indexOf('Error getting docker client:') !== -1) {
    throw new Error('Failed getting docker client')
  }
  if (error.indexOf('Error processing image:') !== -1) {
    throw new Error('Failed processing image:' + targetImage)
  }
}

function getDependencies(analyzerBinaryPath, targetImage) {
  var result;
  return subProcess.execute(
    analyzerBinaryPath,
    buildArgs(targetImage)
  )
  .then(function (output) {
    result = parseAnalysisResults(output);
    return buildTree(
      targetImage, result.type, result.depInfosList, result.osRelease);
  })
  .then(function (package) {
    return {
      package: package,
      packageManager: result.type,
    };
  })
  .catch(function (error) {
    if (typeof error === 'string') {
      debug(`Error while running analyser: '${error}'`);
      handleCommonErrors(error, targetImage);
      errorMsg = error;
      errorMatch = /msg="(.*)"/g.exec(errorMsg)
      if (errorMatch) {
        errorMsg = errorMatch[1]
      }
      throw new Error(errorMsg);
    }

    throw error;
  })
}

function parseAnalysisResults(analysisOut) {
  var analysisJson = JSON.parse(analysisOut);

  var analysisResult = analysisJson.results.filter(function (res) {
    return res.Analysis && res.Analysis.length > 0;
  })[0];

  // TODO: if analysisResult is undefined -
  //  throw a nice error, or report 0 deps

  var depType;
  switch (analysisResult.AnalyzeType) {
    case 'Apt': {
      depType = 'deb';
      break;
    }
    default: {
      depType = analysisResult.AnalyzeType.toLowerCase()
    }
  }

  return {
    type: depType,
    depInfosList: analysisResult.Analysis,
    osRelease: analysisJson.osRelease,
  }
}

function buildTree(targetImage, depType, depInfosList, osRelease) {
  var targetSplit = targetImage.split(':');
  var imageName = targetSplit[0];
  var imageVersion = targetSplit[1] ? targetSplit[1] : 'latest';

  var root = {
    name: imageName,
    version: imageVersion,
    dockerOSRelease: osRelease,
    packageFormatVersion: depType + ':0.0.1',
    dependencies: {},
  };

  var depsMap = depInfosList.reduce(function (acc, depInfo) {
    var name = depInfo.Name;
    acc[name] = depInfo;
    return acc;
  }, {});

  var virtualDepsMap = depInfosList.reduce(function (acc, depInfo) {
    var providesNames = depInfo.Provides || [];
    providesNames.forEach(function (name) {
      acc[name] = depInfo;
    });
    return acc;
  }, {});

  var attachDeps = function (depInfos) {
    depInfos.forEach(function (depInfo) {
      var subtree = buildTreeRecurisve(
        depInfo.Name, new Set(), depsMap, virtualDepsMap);
      if (subtree) {
        root.dependencies[subtree.name] = subtree;
      }
    })
  };

  // attach (as direct deps) pkgs not marked auto-installed:
  var manuallyInstalledDeps = depInfosList.filter(function (depInfo) {
    return !depInfo.AutoInstalled;
  })
  attachDeps(manuallyInstalledDeps);

  // attach (as direct deps) pkgs marked as auto-insatalled,
  //  but not dependant upon:
  var notVisitedDeps = depInfosList.filter(function (depInfo) {
    var depName = depInfo.Name;
    return !(depsMap[depName]._visited);
  })
  attachDeps(notVisitedDeps);

  return root;
}

function buildTreeRecurisve(
    depName, ancestors, depsMap, virtualDepsMap) {
  var depInfo = depsMap[depName] || virtualDepsMap[depName];
  if (!depInfo) {
    return null;
  }

  var fullName = depFullName(depInfo);
  if (ancestors.has(fullName)) {
    return null;
  }

  const tree = {
    name: fullName,
    version: depInfo.Version,
  };

  //TODO: if we do this we can avoid tracking ancestors
  if (depInfo._visited) {
    return tree;
  }
  depInfo._visited = true;

  var newAncestors = (new Set(ancestors)).add(fullName);

  var deps = depInfo.Deps || {};
  Object.keys(deps).forEach(function (depName) {
    var subTree = buildTreeRecurisve(
      depName, newAncestors, depsMap, virtualDepsMap);
    if (subTree) {
      if (!tree.dependencies) {
        tree.dependencies = {};
      }
      tree.dependencies[subTree.name] = subTree;
    }
  });

  return tree;
}

function depFullName(depInfo) {
  var fullName = depInfo.Name;
  if (depInfo.Source) {
    fullName = depInfo.Source + '/' + fullName;
  }
  return fullName;
}

function buildArgs(targetImage) {
  var args = ['analyze', targetImage];
  return args;
}
