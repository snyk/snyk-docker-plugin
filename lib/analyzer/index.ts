import * as Debug from "debug";
import { Docker, DockerOptions } from "../docker";
import * as dockerFile from "../docker-file";
import * as apkAnalyzer from "./apk-analyzer";
import * as aptAnalyzer from "./apt-analyzer";
import * as binariesAnalyzer from "./binaries-analyzer";
import * as imageInspector from "./image-inspector";
import * as osReleaseDetector from "./os-release-detector";
import * as rpmAnalyzer from "./rpm-analyzer";

export { analyze };

const debug = Debug("snyk");

const extractActions = [
  ...aptAnalyzer.APT_PKGPATHS,
  ...apkAnalyzer.APK_PKGPATHS,
  ...osReleaseDetector.OS_VERPATHS,
].map((p) => {
  return {
    name: "txt",
    pattern: p,
  };
});

async function analyze(
  targetImage: string,
  dockerfileAnalysis?: dockerFile.DockerFileAnalysis,
  options?: DockerOptions,
) {
  const docker = new Docker(targetImage, options);

  await docker.scanStaticalyIfNeeded(extractActions);

  const [imageInspection, osRelease] = await Promise.all([
    imageInspector.detect(docker),
    osReleaseDetector.detect(docker, dockerfileAnalysis),
  ]);

  const results = await Promise.all([
    apkAnalyzer.analyze(docker),
    aptAnalyzer.analyze(docker),
    rpmAnalyzer.analyze(docker),
  ]).catch((err) => {
    debug(`Error while running analyzer: '${err.stderr}'`);
    throw new Error("Failed to detect installed OS packages");
  });

  const { installedPackages, pkgManager } = getInstalledPackages(
    results as any[],
  );
  let binaries;
  try {
    binaries = await binariesAnalyzer.analyze(
      targetImage,
      installedPackages,
      pkgManager,
      options,
    );
  } catch (err) {
    debug(`Error while running binaries analyzer: '${err}'`);
    throw new Error("Failed to detect binaries versions");
  }

  return {
    imageId: imageInspection.Id,
    osRelease,
    results,
    binaries,
    imageLayers: imageInspection.RootFS && imageInspection.RootFS.Layers,
  };
}

function getInstalledPackages(
  results: any[],
): { installedPackages: string[]; pkgManager?: string } {
  const dockerAnalysis = results.find((res) => {
    return res.Analysis && res.Analysis.length > 0;
  });

  if (!dockerAnalysis) {
    return { installedPackages: [] };
  }
  const installedPackages = dockerAnalysis.Analysis.map((pkg) => pkg.Name);
  let pkgManager = dockerAnalysis.AnalyzeType;
  if (pkgManager) {
    pkgManager = pkgManager.toLowerCase();
  }
  return { installedPackages, pkgManager };
}
