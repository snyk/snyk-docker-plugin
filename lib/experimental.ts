import * as fs from "fs";
import * as os from "os";
import * as path from "path";

import { pullIfNotLocal } from "./analyzer/image-inspector";
import { Docker } from "./docker";
import { getDockerArchivePath, getImageTransport } from "./image-transport";
import * as staticModule from "./static";
import {
  ImageTransport,
  ImageType,
  PluginResponse,
  StaticAnalysisOptions,
} from "./types";

export async function experimentalAnalysis(
  targetImage: string,
  options: any,
): Promise<PluginResponse> {
  // assume Distroless scanning
  const imageTransport = getImageTransport(targetImage);
  switch (imageTransport) {
    case ImageTransport.DockerArchive:
      // TODO: maybe all the steps in this case can be wrapped in a function
      const imagePath = getDockerArchivePath(targetImage);
      // TODO: validate the file in "imagePath" exists
      const staticAnalysisOptions: StaticAnalysisOptions = {
        imagePath,
        imageType: ImageType.DockerArchive,
        distroless: true,
      };
      return staticModule.analyzeStatically(targetImage, {
        staticAnalysisOptions,
      });

    case ImageTransport.ContainerRegistry:
      return distroless(targetImage, options);

    default:
      throw new Error("Unhandled image transport for image " + targetImage);
  }
}

// experimental flow expected to be merged with the static analysis when ready
export async function distroless(
  targetImage: string,
  options: any,
): Promise<PluginResponse> {
  if (staticModule.isRequestingStaticAnalysis(options)) {
    options.staticAnalysisOptions.distroless = true;
    return staticModule.analyzeStatically(targetImage, options);
  }

  await pullIfNotLocal(targetImage);

  const archiveDir = path.join(os.tmpdir(), "snyk-image-archives");
  createTempDirIfMissing(archiveDir);
  // TODO terrible way to convert slashes to anything else
  // so we don't think it's a directory
  const archiveFileName = `${targetImage.replace(/\//g, "__")}.tar`;
  const archiveFullPath = path.join(archiveDir, archiveFileName);

  // assumption #1: the `docker` binary is available locally
  const docker = new Docker(targetImage);
  await docker.save(targetImage, archiveFullPath);
  try {
    const scanningOptions = {
      staticAnalysisOptions: {
        imagePath: archiveFullPath,
        imageType: ImageType.DockerArchive,
        distroless: true,
      },
    };

    return await staticModule.analyzeStatically(targetImage, scanningOptions);
  } finally {
    fs.unlinkSync(archiveFullPath);
  }
}

function createTempDirIfMissing(archiveDir: string): void {
  try {
    fs.mkdirSync(archiveDir);
  } catch (err) {
    if (err.code !== "EEXIST") {
      throw err;
    }
  }
}
