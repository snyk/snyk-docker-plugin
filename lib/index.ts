const debug = require('debug')('snyk');

import * as analyzer from './analyzer';
import * as subProcess from './sub-process';
import * as dockerFile from './docker-file';

export {
  inspect,
};

function inspect(root: string, targetFile?: string, options?: any) {
  const targetImage = root;
  return Promise.all([
    getRuntime(),
    getDependencies(targetImage),
    dockerFile.getBaseImageName(targetFile),
  ])
    .then((result) => {
      const metadata = {
        name: 'snyk-docker-plugin',
        runtime: result[0],
        packageManager: result[1].packageManager,
        dockerImageId: result[1].imageId,
      };
      const pkg: any = result[1].package;
      pkg.docker = pkg.docker || {};
      pkg.docker.baseImage = result[2];
      return {
        plugin: metadata,
        package: pkg,
      };
    });
}

function getRuntime() {
  return subProcess.execute('docker', ['version'])
    .then((output) => {
      const versionMatch = /Version:\s+(.*)\n/.exec(output);
      if (versionMatch) {
        return 'docker ' + versionMatch[1];
      }
      return undefined;
    });
}

function handleCommonErrors(error, targetImage: string) {
  if (error.indexOf('command not found') !== -1) {
    throw new Error('Snyk docker CLI was not found');
  }
  if (error.indexOf('Cannot connect to the Docker daemon') !== -1) {
    throw new Error('Cannot connect to the Docker daemon. Is the docker'
    + ' daemon running?');
  }
  const ERROR_LOADING_IMAGE_STR = 'Error loading image from docker engine:';
  if (error.indexOf(ERROR_LOADING_IMAGE_STR) !== -1) {
    if (error.indexOf('reference does not exist') !== -1) {
      throw new Error(
        `Docker image was not found locally: ${targetImage}`);
    }
    if (error.indexOf('permission denied while trying to connect') !== -1) {
      let errString = error.split(ERROR_LOADING_IMAGE_STR)[1];
      errString = (errString || '').slice(0, -2); // remove trailing \"
      throw new Error(
        'Permission denied connecting to docker daemon. ' +
        'Please make sure user has the required permissions. ' +
        'Error string: ' + errString);
    }
  }
  if (error.indexOf('Error getting docker client:') !== -1) {
    throw new Error('Failed getting docker client');
  }
  if (error.indexOf('Error processing image:') !== -1) {
    throw new Error('Failed processing image:' + targetImage);
  }
}

function getDependencies(targetImage: string) {
  let result;
  return analyzer.analyze(targetImage)
    .then((output) => {
      result = parseAnalysisResults(output);
      return buildTree(
        targetImage, result.type, result.depInfosList, result.targetOS);
    })
    .then((pkg) => {
      return {
        package: pkg,
        packageManager: result.type,
        imageId: result.imageId,
      };
    })
    .catch((error) => {
      if (typeof error === 'string') {
        debug(`Error while running analyzer: '${error}'`);
        handleCommonErrors(error, targetImage);
        let errorMsg = error;
        const errorMatch = /msg="(.*)"/g.exec(errorMsg);
        if (errorMatch) {
          errorMsg = errorMatch[1];
        }
        throw new Error(errorMsg);
      }
      throw error;
    });
}

function parseAnalysisResults(analysisJson) {
  const analysisResult = analysisJson.results.filter((res) => {
    return res.Analysis && res.Analysis.length > 0;
  })[0];

  if (!analysisResult) {
    throw new Error(
      'Failed to detect a supported Linux package manager (deb/rpm/apk)');
  }

  let depType;
  switch (analysisResult.AnalyzeType) {
    case 'Apt': {
      depType = 'deb';
      break;
    }
    default: {
      depType = analysisResult.AnalyzeType.toLowerCase();
    }
  }

  return {
    imageId: analysisJson.imageId,
    targetOS: analysisJson.osRelease,
    type: depType,
    depInfosList: analysisResult.Analysis,
  };
}

function buildTree(targetImage, depType, depInfosList, targetOS) {
  const targetSplit = targetImage.split(':');
  const imageName = targetSplit[0];
  const imageVersion = targetSplit[1] ? targetSplit[1] : 'latest';

  const root = {
    name: imageName,
    version: imageVersion,
    targetOS,
    packageFormatVersion: depType + ':0.0.1',
    dependencies: {},
  };

  const depsMap = depInfosList.reduce((acc, depInfo) => {
    const name = depInfo.Name;
    acc[name] = depInfo;
    return acc;
  }, {});

  const virtualDepsMap = depInfosList.reduce((acc, depInfo) => {
    const providesNames = depInfo.Provides || [];
    providesNames.forEach((name) => {
      acc[name] = depInfo;
    });
    return acc;
  }, {});

  const depsCounts = {};
  depInfosList.forEach((depInfo) => {
    countDepsRecursive(
      depInfo.Name, new Set(), depsMap, virtualDepsMap, depsCounts);
  });
  const DEP_FREQ_THRESHOLD = 100;
  const tooFrequentDepNames = Object.keys(depsCounts)
    .filter((depName) => {
      return depsCounts[depName] > DEP_FREQ_THRESHOLD;
    });

  const attachDeps = (depInfos) => {
    const depNamesToSkip = new Set(tooFrequentDepNames);
    depInfos.forEach((depInfo) => {
      const subtree = buildTreeRecurisve(
        depInfo.Name, new Set(), depsMap, virtualDepsMap, depNamesToSkip);
      if (subtree) {
        root.dependencies[subtree.name] = subtree;
      }
    });
  };

  // attach (as direct deps) pkgs not marked auto-installed:
  const manuallyInstalledDeps = depInfosList.filter((depInfo) => {
    return !depInfo.AutoInstalled;
  });
  attachDeps(manuallyInstalledDeps);

  // attach (as direct deps) pkgs marked as auto-insatalled,
  //  but not dependant upon:
  const notVisitedDeps = depInfosList.filter((depInfo) => {
    const depName = depInfo.Name;
    return !(depsMap[depName]._visited);
  });
  attachDeps(notVisitedDeps);

  // group all the "too frequest" deps under a meta package:
  if (tooFrequentDepNames.length > 0) {
    const tooFrequentDeps = tooFrequentDepNames.map((name) => {
      return depsMap[name];
    });

    const metaSubtree = {
      name: 'meta-common-packages',
      version: 'meta',
      dependencies: {},
    };

    tooFrequentDeps.forEach((depInfo) => {
      const pkg = {
        name: depFullName(depInfo),
        version: depInfo.Version,
      };
      metaSubtree.dependencies[pkg.name] = pkg;
    });

    root.dependencies[metaSubtree.name] = metaSubtree;
  }

  return root;
}

function buildTreeRecurisve(
  depName, ancestors, depsMap, virtualDepsMap, depNamesToSkip) {
  const depInfo = depsMap[depName] || virtualDepsMap[depName];
  if (!depInfo) {
    return null;
  }

  // "realName" as the argument depName might be a virtual pkg
  const realName = depInfo.Name;
  const fullName = depFullName(depInfo);
  if (ancestors.has(fullName) || depNamesToSkip.has(realName)) {
    return null;
  }

  const tree: {
    name: string;
    version: string;
    dependencies?: any;
  } = {
    name: fullName,
    version: depInfo.Version,
  };

  if (depInfo._visited) {
    return tree;
  }
  depInfo._visited = true;

  const newAncestors = (new Set(ancestors)).add(fullName);

  const deps = depInfo.Deps || {};
  Object.keys(deps).forEach((name) => {
    const subTree = buildTreeRecurisve(
      name, newAncestors, depsMap, virtualDepsMap, depNamesToSkip);
    if (subTree) {
      if (!tree.dependencies) {
        tree.dependencies = {};
      }
      tree.dependencies[subTree.name] = subTree;
    }
  });

  return tree;
}

function countDepsRecursive(
  depName, ancestors, depsMap, virtualDepsMap, depCounts) {
  const depInfo = depsMap[depName] || virtualDepsMap[depName];
  if (!depInfo) {
    return;
  }

  // "realName" as the argument depName might be a virtual pkg
  const realName = depInfo.Name;
  if (ancestors.has(realName)) {
    return;
  }

  depCounts[realName] = (depCounts[realName] || 0) + 1;

  const newAncestors = (new Set(ancestors)).add(realName);
  const deps = depInfo.Deps || {};
  Object.keys(deps).forEach((name) => {
    countDepsRecursive(
      name, newAncestors, depsMap, virtualDepsMap, depCounts);
  });
}

function depFullName(depInfo) {
  let fullName = depInfo.Name;
  if (depInfo.Source) {
    fullName = depInfo.Source + '/' + fullName;
  }
  return fullName;
}
