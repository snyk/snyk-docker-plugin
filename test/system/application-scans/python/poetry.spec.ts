import { scan } from "../../../../lib";
import { getFixture } from "../../../util";

describe("poetry application scan", () => {
  it("should correctly return applications as multiple scan results", async () => {
    const fixturePath = getFixture("docker-archives/docker-save/poetry.tar");
    const imageNameAndTag = `docker-archive:${fixturePath}`;

    const pluginResult = await scan({
      path: imageNameAndTag,
    });

    expect(pluginResult).toMatchSnapshot();
    expect(pluginResult.scanResults).toHaveLength(2);
  });

  it("should handle --exclude-app-vulns with string and boolean value", async () => {
    const fixturePath = getFixture("docker-archives/docker-save/php.tar");
    const imageNameAndTag = `docker-archive:${fixturePath}`;

    const pluginResultExcludeAppVulnsFalseString = await scan({
      path: imageNameAndTag,
      "exclude-app-vulns": "false",
    });

    const pluginResultExcludeAppVulnsTrueString = await scan({
      path: imageNameAndTag,
      "exclude-app-vulns": "true",
    });

    const pluginResultExcludeAppVulnsFalseBoolean = await scan({
      path: imageNameAndTag,
      "exclude-app-vulns": false,
    });

    const pluginResultExcludeAppVulnsTrueBoolean = await scan({
      path: imageNameAndTag,
      "exclude-app-vulns": true,
    });

    expect(pluginResultExcludeAppVulnsFalseString.scanResults).toHaveLength(2);
    expect(pluginResultExcludeAppVulnsFalseBoolean.scanResults).toHaveLength(2);

    expect(pluginResultExcludeAppVulnsTrueString.scanResults).toHaveLength(1);
    expect(pluginResultExcludeAppVulnsTrueBoolean.scanResults).toHaveLength(1);
  });

  it("should handle --collect-application-files", async () => {
    const fixturePath = getFixture(
      "docker-archives/docker-save/poetry-flask.tar",
    );
    const imageNameAndTag = `docker-archive:${fixturePath}`;

    const resultWithoutApplicationFilesFlag = await scan({
      path: imageNameAndTag,
    });
    const resultWithApplicationFilesFlagSetToTrue = await scan({
      path: imageNameAndTag,
      "collect-application-files": "true",
    });

    expect(resultWithoutApplicationFilesFlag.scanResults).toHaveLength(2);
    expect(resultWithApplicationFilesFlagSetToTrue.scanResults).toHaveLength(3);

    const appFiles =
      resultWithApplicationFilesFlagSetToTrue.scanResults[2].facts.find(
        (fact) => fact.type === "applicationFiles",
      )!.data;
    expect(appFiles[0].language).toStrictEqual("python");
    expect(appFiles[0].fileHierarchy).toStrictEqual([{ path: "server.py" }]);
  });
});
