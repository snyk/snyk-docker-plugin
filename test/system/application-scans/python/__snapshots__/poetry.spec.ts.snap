// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`poetry application scan should correctly return applications as multiple scan results 1`] = `
Object {
  "scanResults": Array [
    Object {
      "facts": Array [
        Object {
          "data": Object {
            "graph": Object {
              "nodes": Array [
                Object {
                  "deps": Array [],
                  "nodeId": "root-node",
                  "pkgId": "docker-image|poetry.tar@",
                },
              ],
              "rootNodeId": "root-node",
            },
            "pkgManager": Object {
              "name": "linux",
              "repositories": Array [
                Object {
                  "alias": "unknown:0.0",
                },
              ],
            },
            "pkgs": Array [
              Object {
                "id": "docker-image|poetry.tar@",
                "info": Object {
                  "name": "docker-image|poetry.tar",
                  "version": undefined,
                },
              },
            ],
            "schemaVersion": "1.3.0",
          },
          "type": "depGraph",
        },
        Object {
          "data": "sha256:83878e4a19e2c44d6efc14d3aa75b0cda05a44927575480cf3cef8889efa26f9",
          "type": "imageId",
        },
        Object {
          "data": Array [
            "a5ceb4cf1de62292b28ff91de7c88878fe655a60cfb0042d5b0a7db7cc998d97/layer.tar",
          ],
          "type": "imageLayers",
        },
        Object {
          "data": "2022-08-04T13:17:10.411032291Z",
          "type": "imageCreationTime",
        },
        Object {
          "data": Array [
            "sha256:1a94d0e60a13af7e273a428cb3f439f1f8872c806a70c41895bfd95de875c4c4",
          ],
          "type": "rootFs",
        },
      ],
      "identity": Object {
        "args": Object {
          "platform": "linux/amd64",
        },
        "type": "linux",
      },
      "target": Object {
        "image": "docker-image|poetry.tar",
      },
    },
    Object {
      "facts": Array [
        Object {
          "data": Object {
            "graph": Object {
              "nodes": Array [
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "six",
                    },
                    Object {
                      "nodeId": "flask",
                    },
                    Object {
                      "nodeId": "rpc.py",
                    },
                  ],
                  "nodeId": "root-node",
                  "pkgId": "myPkg@1.42.2",
                },
                Object {
                  "deps": Array [],
                  "info": Object {
                    "labels": Object {
                      "scope": "prod",
                    },
                  },
                  "nodeId": "six",
                  "pkgId": "six@1.16.0",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "click",
                    },
                    Object {
                      "nodeId": "importlib-metadata",
                    },
                    Object {
                      "nodeId": "itsdangerous",
                    },
                    Object {
                      "nodeId": "jinja2",
                    },
                    Object {
                      "nodeId": "werkzeug",
                    },
                  ],
                  "info": Object {
                    "labels": Object {
                      "scope": "prod",
                    },
                  },
                  "nodeId": "flask",
                  "pkgId": "flask@2.2.0",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "colorama",
                    },
                    Object {
                      "nodeId": "importlib-metadata",
                    },
                  ],
                  "info": Object {
                    "labels": Object {
                      "scope": "prod",
                    },
                  },
                  "nodeId": "click",
                  "pkgId": "click@8.1.3",
                },
                Object {
                  "deps": Array [],
                  "info": Object {
                    "labels": Object {
                      "scope": "prod",
                    },
                  },
                  "nodeId": "colorama",
                  "pkgId": "colorama@0.4.5",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "typing-extensions",
                    },
                    Object {
                      "nodeId": "zipp",
                    },
                  ],
                  "info": Object {
                    "labels": Object {
                      "scope": "prod",
                    },
                  },
                  "nodeId": "importlib-metadata",
                  "pkgId": "importlib-metadata@4.12.0",
                },
                Object {
                  "deps": Array [],
                  "info": Object {
                    "labels": Object {
                      "scope": "prod",
                    },
                  },
                  "nodeId": "typing-extensions",
                  "pkgId": "typing-extensions@3.10.0.2",
                },
                Object {
                  "deps": Array [],
                  "info": Object {
                    "labels": Object {
                      "scope": "prod",
                    },
                  },
                  "nodeId": "zipp",
                  "pkgId": "zipp@3.8.1",
                },
                Object {
                  "deps": Array [],
                  "info": Object {
                    "labels": Object {
                      "scope": "prod",
                    },
                  },
                  "nodeId": "itsdangerous",
                  "pkgId": "itsdangerous@2.1.2",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "markupsafe",
                    },
                  ],
                  "info": Object {
                    "labels": Object {
                      "pkgIdProvenance": "Jinja2@3.1.2",
                      "scope": "prod",
                    },
                  },
                  "nodeId": "jinja2",
                  "pkgId": "jinja2@3.1.2",
                },
                Object {
                  "deps": Array [],
                  "info": Object {
                    "labels": Object {
                      "pkgIdProvenance": "MarkupSafe@2.1.1",
                      "scope": "prod",
                    },
                  },
                  "nodeId": "markupsafe",
                  "pkgId": "markupsafe@2.1.1",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "markupsafe",
                    },
                  ],
                  "info": Object {
                    "labels": Object {
                      "pkgIdProvenance": "Werkzeug@2.2.1",
                      "scope": "prod",
                    },
                  },
                  "nodeId": "werkzeug",
                  "pkgId": "werkzeug@2.2.1",
                },
                Object {
                  "deps": Array [
                    Object {
                      "nodeId": "typing-extensions",
                    },
                  ],
                  "info": Object {
                    "labels": Object {
                      "scope": "prod",
                    },
                  },
                  "nodeId": "rpc.py",
                  "pkgId": "rpc.py@0.4.2",
                },
              ],
              "rootNodeId": "root-node",
            },
            "pkgManager": Object {
              "name": "poetry",
            },
            "pkgs": Array [
              Object {
                "id": "myPkg@1.42.2",
                "info": Object {
                  "name": "myPkg",
                  "version": "1.42.2",
                },
              },
              Object {
                "id": "six@1.16.0",
                "info": Object {
                  "name": "six",
                  "version": "1.16.0",
                },
              },
              Object {
                "id": "flask@2.2.0",
                "info": Object {
                  "name": "flask",
                  "version": "2.2.0",
                },
              },
              Object {
                "id": "click@8.1.3",
                "info": Object {
                  "name": "click",
                  "version": "8.1.3",
                },
              },
              Object {
                "id": "colorama@0.4.5",
                "info": Object {
                  "name": "colorama",
                  "version": "0.4.5",
                },
              },
              Object {
                "id": "importlib-metadata@4.12.0",
                "info": Object {
                  "name": "importlib-metadata",
                  "version": "4.12.0",
                },
              },
              Object {
                "id": "typing-extensions@3.10.0.2",
                "info": Object {
                  "name": "typing-extensions",
                  "version": "3.10.0.2",
                },
              },
              Object {
                "id": "zipp@3.8.1",
                "info": Object {
                  "name": "zipp",
                  "version": "3.8.1",
                },
              },
              Object {
                "id": "itsdangerous@2.1.2",
                "info": Object {
                  "name": "itsdangerous",
                  "version": "2.1.2",
                },
              },
              Object {
                "id": "jinja2@3.1.2",
                "info": Object {
                  "name": "jinja2",
                  "version": "3.1.2",
                },
              },
              Object {
                "id": "markupsafe@2.1.1",
                "info": Object {
                  "name": "markupsafe",
                  "version": "2.1.1",
                },
              },
              Object {
                "id": "werkzeug@2.2.1",
                "info": Object {
                  "name": "werkzeug",
                  "version": "2.2.1",
                },
              },
              Object {
                "id": "rpc.py@0.4.2",
                "info": Object {
                  "name": "rpc.py",
                  "version": "0.4.2",
                },
              },
            ],
            "schemaVersion": "1.3.0",
          },
          "type": "depGraph",
        },
        Object {
          "data": Array [
            "pyproject.toml",
            "poetry.lock",
          ],
          "type": "testedFiles",
        },
        Object {
          "data": "sha256:83878e4a19e2c44d6efc14d3aa75b0cda05a44927575480cf3cef8889efa26f9",
          "type": "imageId",
        },
      ],
      "identity": Object {
        "targetFile": "/app/pyproject.toml",
        "type": "poetry",
      },
      "target": Object {
        "image": "docker-image|poetry.tar",
      },
    },
  ],
}
`;
